File: src/main/java/com/example/chatService/listener/ParticipantForcedExitListener.java
package com.example.chatService.listener;

import com.example.chatService.event.ParticipantForcedExitEvent;
import com.example.chatService.service.RoomParticipantService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

import java.util.Map;

@Component
@RequiredArgsConstructor
@Slf4j
public class ParticipantForcedExitListener {
    private final SimpMessagingTemplate messagingTemplate;

    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    public void handle(ParticipantForcedExitEvent event) {

        String roomId = event.getRoomId();
        String userId = event.getUserId();
        String reason = event.getReason();

        messagingTemplate.convertAndSendToUser(
                userId,
                "/queue/room-force-exit",
                Map.of(
                        "roomId", roomId,
                        "reason", reason
                )
        );
        log.info(
                "[FORCED_EXIT] roomId={}, userId={}, reason={}",
                roomId, userId, reason);
    }
}

File: src/main/java/com/example/chatService/listener/ParticipantForcedExitListener.kt
package com.example.chatService.listener

import com.example.chatService.event.ParticipantForcedExitEvent
import org.slf4j.LoggerFactory
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Component
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Component
class ParticipantForcedExitListener(
    private val messagingTemplate: SimpMessagingTemplate,
) {
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    fun handle(event: ParticipantForcedExitEvent) {
        messagingTemplate.convertAndSendToUser(
            event.userId,
            "/queue/room-force-exit",
            mapOf("roomId" to event.roomId, "reason" to event.reason),
        )
        log.info("[FORCED_EXIT] roomId={}, userId={}, reason={}", event.roomId, event.userId, event.reason)
    }

    companion object {
        private val log = LoggerFactory.getLogger(ParticipantForcedExitListener::class.java)
    }
}
